### PR TITLE
fix(theme-ui): add variants to the Theme interface

### PR DESCRIPTION
**Pull request by wKovacs64**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://theme-ui.com/theme-spec
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## theme-ui ([<kbd>unpkg</kbd>](https://unpkg.com/browse/theme-ui@latest/))
was missing the following properties:
1. components
1. get
1. BaseStyles
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 313d14d4fb94911d037551e75c6982528bf6c544
</p>

**Comment by wKovacs64** [comment]
This can probably be simplified somehow by making a type alias for all the variant groups, e.g.

```ts
type VariantGroups =
    | 'grids'
    | 'buttons'
    | 'text'
    | 'links'
    | 'images'
    | 'cards'
    | 'layout'
    | 'forms'
    | 'badges'
    | 'alerts'
    | 'messages';
```

...but I couldn't figure out how to use them as computed properties in the `Theme` interface without changing it to a type alias and some other junk. 🤷‍♂️ 

Also not entirely sure if this belongs here or in the `styled-system` types, but it looked as though SS has its own thing (e.g. `buttons`, `colorStyles`, `textStyles`) and I don't know much about SS itself.

**Comment by wKovacs64** [comment]
Just realized the [comment block](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45042/files#diff-e819233ce58be0eb777ba05411226464R116-R125) about variants that I put in is only applying to the first one (`grids`), which is how those comments work but not ideal. I could duplicate the comment 10 more times to apply to each variant (yuck?) unless someone knows how to turn those into a computed property/loop so they inherit the comment (does that even happen?). Or remove it all together (not great).

**Comment by wKovacs64** [comment]
Any thoughts on this, @hasparus?

**Comment by hasparus** [comment]
> I could duplicate the comment 10 more times to apply to each variant (yuck?) **unless someone knows how to turn those into a computed property/loop so they inherit the comment (does that even happen?)**. 

I tried. I don't think it's possible.

> Also not entirely sure if this belongs here or in the styled-system types, but it looked as though SS has its own thing (e.g. buttons, colorStyles, textStyles) and I don't know much about SS itself.

theme-ui types probably shouldn't depend on styled-system types.
We're working on TypeScript rewrite on theme-ui/master (now published as 0.4-alpha).
We copied the types and we adapt them, because that's what happened with styled-system JS code too.

> This can probably be simplified somehow by making a type alias for all the variant groups, e.g.

This works:

```
type VariantGroups =
    | 'grids'
    | 'buttons'
    | 'text'
    | 'links'
    | 'images'
    | 'cards'
    | 'layout'
    | 'forms'
    | 'badges'
    | 'alerts'
    | 'messages';


type ThemeVariants = {
  [V in VariantGroups]: unknown
}

interface Theme extends ThemeVariants {
  otherStuff: unknown
}
```
[TS Playground](https://www.typescriptlang.org/v2/en/play?#code/C4TwDgpgBAaghgJwJZwHbAOIIPYFcwDOUAvAFBQVQA+UA5AObIAmBt5lNtARrsMNqlbsKnYBAAewNpWp0ANklQBrITM5IAtnHoRVHOgGNELaftpy4IPFOGzaAM2wINekXS5wmO13bhyICMA+nBq6BNq6tADcpLGgkFAAKgAWEKHwyGhBJFAA3uwA2jBQirCIKOhYeIQAugBcUNhcAFYQBsCkAL6ximII9nAG0Clp0BJiqCxJqenlWUT5XUA)

Though I don't dislike listing them explicitly.

**Review by hasparus — APPROVED** [review]

**Comment by wKovacs64** [comment]
> This works:

Yeah that's how far I got too, but since TS doesn't support multiple `extends` and `Theme` already extends a subset of `StyledSystemTheme`, I had to add even more type aliases and it just felt... not worth it. I'm cool listing them explicitly if you are.

I'm going to duplicate the comment, though - as gross as it feels, I think it's a nice DX for the consumer to get the TSDoc tooltip for each variant group.

**Comment by wKovacs64** [comment]
Ah, cool, looks like the lint rules have changed since my original commit so now CI is failing on an unrelated line. Guess I'll change that in this PR to get CI to pass, even though it's unrelated. ¯\\\_(ツ)\_/¯

**Comment by hasparus** [comment]
> Yeah that's how far I got too, **but since TS doesn't support multiple extends** and Theme already extends a subset of StyledSystemTheme...


I've got a gift for you.
JavaScript doesn't support multiple inheritance, what's fair because it usualy leads to a world of pain.

But `extends` keyword for interfaces is basically a nicer type intersection than `&` operator :D
(It detects conflicts on properties.)
I'll admit reusing this keyword is fair, it makes sense. It's certainly better than `as const` and much better than `const = 0` from C++.

```ts
interface A {
  a: 2
};

interface B {
  b: 3
}

interface C {
  c: 4
}

interface D extends A, B, C {}

type KeyofD = keyof D;
```
[TS Playground](https://www.typescriptlang.org/v2/en/play?#code/JYOwLgpgTgZghgYwgAgILIN4ChnLgLmQCYsBfAbiy1ElkRQCFMdkAjQgZjKpuniWQBhZrgSEALN2rg+9ZABFkEAB6QQAEwDOaADTIGe4RlJUwATwAOKANIQzAexiKAvMgDWdxwvJA)

---

However, listing them and adding a tsdoc to each one is probably a solid sacrifice for greater good.

BTW Would you like to PR theme-ui to add these variants?
https://github.com/system-ui/theme-ui/blob/master/packages/css/src/types.ts#L573

**Comment by wKovacs64** [comment]
> I've got a gift for you.

😮 Wow, TIL. Thank you! ❤️

> BTW Would you like to PR theme-ui to add these variants?

Sure thing.

**Comment by wKovacs64** [comment]
Not sure what (if anything) I can do about Travis failing:

```
Error: Cannot find module '/home/travis/.dts/typescript-installs/4.0/node_modules/typescript'
```

GitHub Actions and Azure seem fine. 🤷

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## theme-ui ([<kbd>unpkg</kbd>](https://unpkg.com/browse/theme-ui@latest/))
was missing the following properties:
1. components
1. get
1. BaseStyles
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against cf779e978215dc9f65801a07c02e2cea4a2f33e1
</p>

**Comment by wKovacs64** [comment]
I guess this needs another review when you have time, @hasparus. Pushed the same changes I made to the comments in `theme-ui`.

File: types/theme-ui/index.d.ts [hunk 1]
```diff
@@ -75,7 +75,7 @@ export type ColorMode = {
     accent?: CSS.ColorProperty;
 };
 
-export interface Theme extends Omit<StyledSystemTheme, 'colors'> {
+export interface Theme extends Omit<Omit<StyledSystemTheme, 'colors'>, 'buttons'> {
```
Comment (hasparus) [inline types/theme-ui/index.d.ts:78]: What do you think about this? 👇

```suggestion
export interface Theme extends Omit<StyledSystemTheme, 'colors' | 'buttons'> {
```

**Comment by hasparus** [comment]
Looks good to me.

File: types/theme-ui/index.d.ts (hunk 1 above)
Comment (wKovacs64) [reply types/theme-ui/index.d.ts:78]: Yeah that's better. My TS inexperience is showing.

**Review by sbardian — APPROVED** [review]

**Pull request merged**